### PR TITLE
Don't show help for 'brew <command> help', help-related improvements

### PR DESCRIPTION
**Pull request by UniqMartin**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/.github/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes?
- [x] Have you successfully ran `brew tests` with your changes locally?

----

Summary:

- Let the `help` command handle most of the help display logic (previously in `brew.rb`).
  - Eliminate now unused helper methods `ARGV.usage` and `Homebrew.help_s`.
- Stop interpreting `brew <command> [<arguments>] help` as a request for help. This is hardly useful and prevents commands from receiving a literal `help` as a named argument. (`brew help <command>` and option-style help flags like `--help` continue to be interpreted as usual.)

### CODE REVIEW

File: Library/Homebrew/cmd/help.rb [hunk 1]
```diff
@@ -32,14 +32,34 @@
 # NOTE The reason the string is at the top is so 25 lines is easy to measure!
 
 module Homebrew
-  def help
-    puts HOMEBREW_HELP
-  end
+  def help(cmd = nil, empty_argv = false)
```
Comment (mikemcquaid) [inline Library/Homebrew/cmd/help.rb]: I'd personally prefer `options={}` with an `[:empty_argv]` just for readability; boolean arguments are often hard to parse from their calling side otherwise.

File: Library/brew.rb [hunk 2]
```diff
@@ -39,7 +39,7 @@ def require?(path)
     if help_flag && cmd
       break
     elsif help_flag_list.include? arg
-      help_flag = true
+      help_flag = true unless cmd && arg == "help" # No help when `<cmd> help`.
```
Comment (mikemcquaid) [inline Library/brew.rb]: I find the `unless .. &&` really hard to parse. Maybe split this into a few lines/statements?

**Comment by mikemcquaid** [comment]
I like this 👍 but a few nits.

File: Library/Homebrew/cmd/help.rb (hunk 1 above)
Comment (UniqMartin) [inline Library/Homebrew/cmd/help.rb]: This will make the call site look `Homebrew.help cmd, :empty_argv => empty_argv`. If this doesn't bother you, I'm happy to make the change according to your suggestion.

Comment (mikemcquaid) [inline Library/Homebrew/cmd/help.rb]: If you suspect that'll be the only call-site: feel free to leave as-is :

Comment (UniqMartin) [inline Library/Homebrew/cmd/help.rb]: Very likely. I presume the call site will always be either `Homebrew.help cmd, empty_argv` or just `Homebrew.help`, i.e. there will never be a literal second argument.

**Comment by UniqMartin** [comment]
Updated and incorporated the code comments.

**Comment by apjanke** [comment]
:+1: The existing complicated `help` behavior puts a knot in option handling, and this helps untangle it.

**Comment by apjanke** [comment]
Nitpick though: maybe the PR could be more specific than "various improvements", especially since this changes user-facing behavior, so that people browsing the PR list would know to pay more attention to it.

**Comment by UniqMartin** [comment]
> Nitpick though: maybe the PR could be more specific than "various improvements", especially since this changes user-facing behavior, so that people browsing the PR list would know to pay more attention to it.

Thanks for this input! I've to admit that I didn't put much thought into the PR title. Maybe the core problem is that I lumped together several somewhat related changes because I felt they didn't deserve their own independent PRs. Anyway, I retroactively changed the title and I hope it has improved.

**Comment by apjanke** [comment]
Yep!

This was just on my mind after reading some of the user feedback and discussion after the core/formula split. My takeaway from that is that the PR queue is a major form of communicating recent and upcoming changes to users. And since most don't have time to read everything, at best they'll just be skimming the list and seeing the subject lines, and diving in to the few whose subjects look relevant to them.
